harness/determined/common/storage/shared.py

```diff
@@ -169,18 +170,51 @@ def restore_path(
             raise errors.CheckpointNotFound(f"Did not find checkpoint {src} in shared_fs storage")
         yield pathlib.Path(storage_dir)
 
-    def delete(self, tgt: str) -> None:
+    def delete(self, tgt: str, globs: List[str]) -> Dict[str, int]:
         """
         Delete the stored data from persistent storage.
         """
         storage_dir = os.path.join(self._base_path, tgt)
 
         if not os.path.exists(storage_dir):
             logging.info(f"Storage directory does not exist: {storage_dir}")
-            return
+            return {}
         if not os.path.isdir(storage_dir):
             raise errors.CheckpointNotFound(f"Storage path is not a directory: {storage_dir}")
-        util.rmtree_nfs_safe(storage_dir, ignore_errors=False)
+
+        # Optimize for the common case here. No need to iterate through files.
+        if "**/*" in globs:
+            util.rmtree_nfs_safe(storage_dir, ignore_errors=False)
+            return {}
+
+        to_delete_dirs = {}
+        to_delete_files = {}
+        for file_glob in globs:
+            for path in glob.glob(f"{storage_dir}/{file_glob}", recursive=True):
+                if os.path.isfile(path):
+                    to_delete_files[path] = True
+                elif os.path.isdir(path):
+                    to_delete_dirs[path] = True
+                # TODO should we delete system links too?
+
+        # Delete files first then delete paths.
+        for path in to_delete_files:
+            os.remove(path)  # TODO do we need an os.remove safe here?
+        for path in to_delete_dirs:
+            util.rmtree_nfs_safe(path, ignore_errors=False)
+
+        resources = self._list_directory(storage_dir)
+
+        # TODO we upload this "metadata.json" file in checkpoints sometimes AFTER
+        # we calculate resources so resources won't contain it.
+        # This actually might be an issue if we sometimes have it in resources?
+        # todo this might be an issue.
+        if "metadata.json" in resources:
```

Comment:
this and other `metadata.json` hack is needed since we often create this file after we calculate and store checkpoint resources in the database. 

This causes us when just checking if the checkpoint matches the database state to think that it always doesn't.

Not sure if this handling of it is fine.